### ISSUE TITLE
can i add more features?

### ISSUE THREAD

**Issue by ILIKEBANANANANANA** [rakesh-i/ESP32-Autonomous-car#1]
Hi,
I was planning on building this project with just one question: can I add object detection for ex: it can detect other vehicles or traffic lights?
If yes then will I need to just train it more or will I have to modify the code?
Regards

**Comment by rakesh-i** [comment]
You need to modify the code. The ai model only classifies the images into left right and straight. You can take the same ai model and train it with detecting cars and traffic light and run it parallel to the road detection model. But remember it can classify only 3 objects at a time in current state.

**Comment by ILIKEBANANANANANA** [comment]
Oh thanks, can you send me some references of the code?

**Comment by rakesh-i** [comment]
reference to what?  Everything you need is already there. You can create the logic you want. I'm busy with other projects so i can't put my time into this project.

**Comment by ILIKEBANANANANANA** [comment]
Ok, I will do that.

**Comment by ILIKEBANANANANANA** [comment]
I am confused about which code files will I have to modify the code?

**Comment by ILIKEBANANANANANA** [comment]
i have made some modifications to the vision codefile and this is how it looks:

import cv2
import numpy as np
import tensorflow as tf
import time

# Load the pre-trained object detection model
model = tf.keras.models.load_model('model.h5')

# Load the pre-trained traffic light detection model
traffic_light_model = tf.keras.models.load_model('traffic_light_model.h5')

# Set up the camera capture
cap = cv2.VideoCapture(0)
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

while True:
    # Read a frame from the camera
    ret, frame = cap.read()

    # Preprocess the object detection frame
    object_detection_image = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    object_detection_image = cv2.resize(object_detection_image, (224, 224))
    object_detection_image = object_detection_image / 255.0
    object_detection_image = np.expand_dims(object_detection_image, axis=0)

    # Run the object detection model on the frame
    start_time = time.time()
    object_predictions = model.predict(object_detection_image)
    elapsed_time = time.time() - start_time
    print('Elapsed time for object detection: {:.3f} seconds'.format(elapsed_time))

    # Draw the bounding boxes on the detected objects
    for prediction in object_predictions[0]:
        score = prediction[0]
        if score > 0.5:
            x1 = int(prediction[1] * 640)
            y1 = int(prediction[2] * 480)
            x2 = int(prediction[3] * 640)
            y2 = int(prediction[4] * 480)
            label = 'Object {:.2f}'.format(score)
            cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)
            cv2.putText(frame, label, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

    # Preprocess the traffic light detection frame
    traffic_light_detection_image = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    traffic_light_detection_image = cv2.resize(traffic_light_detection_image, (32, 32))
    traffic_light_detection_image = traffic_light_detection_image / 255.0
    traffic_light_detection_image = np.expand_dims(traffic_light_detection_image, axis=0)

    # Run the traffic light detection model on the frame
    start_time = time.time()
    traffic_light_predictions = traffic_light_model.predict(traffic_light_detection_image)
    elapsed_time = time.time() - start_time
    print('Elapsed time for traffic light detection: {:.3f} seconds'.format(elapsed_time))

    # Check if a traffic light is detected and display the label
    traffic_light_prediction = np.argmax(traffic_light_predictions[0])
    traffic_light_label = ''
    if traffic_light_prediction == 0:
        traffic_light_label = 'Traffic Light: Green'
    elif traffic_light_prediction == 1:
        traffic_light_label = 'Traffic Light: Red'
    elif traffic_light_prediction == 2:
        traffic_light_label = 'Traffic Light: Yellow'
    cv2.putText(frame, traffic_light_label, (10, 20), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 2)

    # Display the resulting frame
    cv2.imshow('Object and Traffic Light Detection', frame)


    # Exit if the 'q' key is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the camera and close all windows
cap.release()
cv2.destroyAllWindows()

**Comment by ILIKEBANANANANANA** [comment]
is this enough or do I need to do more modifications in other code files?

**Comment by ILIKEBANANANANANA** [comment]
thanks in advance

**Comment by rakesh-i** [comment]
I really want to help but I'm too busy with the work. I hope you can work it out. Sorry :(

**Comment by ILIKEBANANANANANA** [comment]
No problem, I understand. Your work is more important than mine I am just
trying to replicate it.But once you are free though,please give me some
guidance I can really use it,take as much time as you need no worries or
hurries :)
